core/src/cell.rs

```diff
@@ -54,32 +56,81 @@ pub struct ResolvedTransaction {
 
 pub trait CellProvider {
     fn cell(&self, out_point: &OutPoint) -> CellStatus;
+}
+
+pub struct OverlayCellProvider<'a, O, CP> {
+    overlay: &'a O,
+    cell_provider: &'a CP,
+}
+
+impl<'a, O, CP> OverlayCellProvider<'a, O, CP> {
+    pub fn new(overlay: &'a O, cell_provider: &'a CP) -> Self {
+        OverlayCellProvider {
+            overlay,
+            cell_provider,
+        }
+    }
+}
+
+impl<'a, O, CP> CellProvider for OverlayCellProvider<'a, O, CP>
+where
+    O: CellProvider,
+    CP: CellProvider,
+{
+    fn cell(&self, out_point: &OutPoint) -> CellStatus {
+        match self.overlay.cell(out_point) {
+            CellStatus::Live(co) => CellStatus::Live(co),
+            CellStatus::Dead => CellStatus::Dead,
+            CellStatus::Unknown => self.cell_provider.cell(out_point),
+        }
+    }
+}
+
+pub struct BlockCellProvider<'a> {
+    output_indexs: FnvHashMap<H256, usize>,
+    block: &'a Block,
+}
 
-    fn cell_at<F: Fn(&OutPoint) -> Option<bool>>(
-        &self,
-        _out_point: &OutPoint,
-        _is_dead: F,
-    ) -> CellStatus {
-        unreachable!()
+impl<'a> BlockCellProvider<'a> {
+    pub fn new(block: &'a Block) -> Self {
+        let mut output_indexs = FnvHashMap::default();
+        for (i, tx) in block.commit_transactions().iter().enumerate() {
+            output_indexs.insert(tx.hash(), i);
+        }
+        Self {
```

Comment:
It would be better if us  `with_capacity` to avoid reallocating. 
Moreover,  use `from_iter`

```
use std::iter::FromIterator;
FnvHashMap::from_iter(
   block.commit_transactions().iter().enumerate().map(|(idx, tx)| (tx.hash(), idx))
)
```

Reply:
Resolved, thanks